librad/src/meta/entity.rs

```diff
@@ -581,35 +654,41 @@ where
     where
         ST: Clone,
     {
-        let mut keys = self.keys().iter().cloned().collect::<HashSet<_>>();
-        let mut users = self.certifiers().iter().cloned().collect::<HashSet<_>>();
-
         if self.revision == 1 && (self.parent_hash.is_some() || self.root_hash != self.hash) {
             // TODO: define a better error if `self.parent_hash.is_some()`
             // (should be "revision 1 cannot have a parent hash")
             return Err(Error::InvalidRootHash);
         }
 
-        for (k, s) in self.signatures() {
-            match self.check_signature(k, &s.by, &s.sig, resolver).await {
-                Err(e) => return Err(e),
-                Ok(_signed) => {},
-            };
-
-            match &s.by {
-                Signatory::OwnedKey => {
-                    keys.remove(k);
-                },
-                Signatory::User(user) => {
-                    users.remove(&user);
+        for (k, s) in self.keys.iter() {
+            match s {
+                Some(sig) => {
+                    if !sig.verify(k, self.hash(), self.revision) {
+                        return Err(Error::SignatureVerificationFailed(k.to_owned()));
+                    }
                 },
+                None => return Err(Error::SignatureMissing),
             }
         }
-        if keys.is_empty() && users.is_empty() {
-            Ok(self.with_status::<Signed>())
-        } else {
-            Err(Error::SignatureMissing)
+
+        for (urn, s) in self.certifiers.iter() {
+            match s {
+                Some(sig) => {
+                    if !sig.verify(self.hash()) {
+                        return Err(Error::SignatureVerificationFailed(sig.key.to_owned()));
+                    }
+                    resolver
```

Comment:
It terrifies me that we could remove this and the code would still compile without any worry. There's a disconnect between the `urn` in the `sig` and the verification.